net/Socket.hpp

```diff
@@ -49,36 +49,43 @@ namespace Poco
 }
 
 class Socket;
+class SocketPoll;
 
 /// Helper to allow us to easily defer the movement of a socket
 /// between polls to clarify thread ownership.
 class SocketDisposition final
 {
-    typedef std::function<void(const std::shared_ptr<Socket> &)> MoveFunction;
-    enum class Type { CONTINUE, CLOSED, MOVE };
-
-    Type _disposition;
-    MoveFunction _socketMove;
-    std::shared_ptr<Socket> _socket;
+    enum class Type { CONTINUE, CLOSED, MOVE, TRANSFER };
 
 public:
+    typedef std::function<void(const std::shared_ptr<Socket> &)> MoveFunction;
+
     SocketDisposition(const std::shared_ptr<Socket> &socket) :
         _disposition(Type::CONTINUE),
+        _toPoll(nullptr),
         _socket(socket)
     {}
     ~SocketDisposition()
     {
         assert (!_socketMove);
     }
-    void setMove()
-    {
-        _disposition = Type::MOVE;
-    }
+    // not the method you want.
     void setMove(MoveFunction moveFn)
     {
         _socketMove = std::move(moveFn);
         _disposition = Type::MOVE;
     }
+    /** move, correctly change ownership of and insert into a new poll.
+     * @transferFn is called as a callback inside the new poll, which
+     * has its thread started if it is not already.
+     * @moveFn can be used for optional setup.
+     */
+    void setTransfer(SocketPoll &poll, MoveFunction transferFn)
```

Comment:
We should really start passing around shared_ptr<SocketPoll>, other cases exist too. Naked references make me nervous.